### PR TITLE
Add ExceptionHandled to ExceptionContext

### LINKED ISSUE #4700: Add ExceptionHandled to ExceptionContext
https://msdn.microsoft.com/en-us/library/system.web.mvc.exceptioncontext.exceptionhandled(v=vs.118).aspx#P:System.Web.Mvc.ExceptionContext.ExceptionHandled is missing from `ExceptionContext`. Our other 'after' context classes have it.

We should add this since this is the place you're most likely to use it.

### PR DESCRIPTION
**Pull request by ryanbrandenburg**
Fixes #4700.

@rynowak please check me extra well since I'm not sure I correctly interpreted the behavior/functionality of this area.

### CODE REVIEW

**Comment by rynowak** [comment]
> please check me extra well since I'm not sure I correctly interpreted the behavior/functionality of this area.

hint: I'd suggest first investigating MVC5's behavior. We found that we were missing this from the API surface of previous versions.

File: src/Microsoft.AspNetCore.Mvc.Abstractions/Filters/ExceptionContext.cs [hunk 1]
```diff
@@ -69,6 +69,11 @@ public virtual ExceptionDispatchInfo ExceptionDispatchInfo
         }
 
         /// <summary>
+        /// Gets or sets an indication that the <see cref="Exception"/> has been handled.
+        /// </summary>
+        public virtual bool ExceptionHandled { get; set; } = false;
```
Comment (javiercn) [inline src/Microsoft.AspNetCore.Mvc.Abstractions/Filters/ExceptionContext.cs]: Why virtual? I don't think there's an easy way to replace the exception context with your own, and being get set I don't see how making it virtual would improve unit testing which is the other scenario I would consider. @rynowak am I missing something?

Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Abstractions/Filters/ExceptionContext.cs]: All of the properties on all of these classes are virtual for mocking support

**Comment by ryanbrandenburg** [comment]
@rynowak after reviewing how we did this in MVC5 I've made an update. I believe this would be a breaking change for any exception filters that currently handle exceptions as they'd need to start marking `ExceptionHandled` as true.

**Comment by rynowak** [comment]
Do you want to explain what the delta in behavior is and why this is terrifying you? I need some deets.

**Comment by ryanbrandenburg** [comment]
:up::date:

The major change from RC2 here (other than just adding the property) is that we set ExceptionHandled to true when handling an exception, rather than nulling the exception. This is consistent with the behavior I observed in MVC5. One thing that has me a bit confused is that yesterday in person I thought I remembered you saying that user code wouldn't be handling this interaction (previously nulling `.Exception`, now setting `.ExceptionHandled` true) but if we look at https://github.com/aspnet/Mvc/blob/dev/test/WebSites/FiltersWebSite/Filters/ShortCircuitExceptionFilter.cs it seems like use defined filters are the ONLY place where ExceptionHandled would get set. Am I mis-remembering that or  missing something else here?

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvoker.cs [hunk 2]
```diff
@@ -422,7 +422,7 @@ private Task InvokeAllExceptionFiltersAsync()
 
                     _diagnosticSource.AfterOnExceptionAsync(_exceptionContext, current.FilterAsync);
 
-                    if (_exceptionContext.Exception == null)
+                    if (_exceptionContext.ExceptionHandled)
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvoker.cs]: Why is this condition different from the other ones? Think what would happen if you set the `.Exception` to null. Would other exception filters run?

File: test/Microsoft.AspNetCore.Mvc.Core.Test/Internal/ControllerActionInvokerTest.cs [hunk 3]
```diff
@@ -155,7 +155,7 @@ public class ControllerActionInvokerTest
                 .Callback<ExceptionContext>(context =>
                 {
                     filter2.ToString();
-                    context.Exception = null;
+                    context.ExceptionHandled = true;
```
Comment (rynowak) [inline test/Microsoft.AspNetCore.Mvc.Core.Test/Internal/ControllerActionInvokerTest.cs]: We should continue to support short circuiting by setting the exception to null. There should be tests for both of these cases

**Comment by rynowak** [comment]
:watch:

**Comment by ryanbrandenburg** [comment]
:up::date:

File: test/Microsoft.AspNetCore.Mvc.Core.Test/Internal/ControllerActionInvokerTest.cs [hunk 4]
```diff
@@ -171,12 +171,39 @@ public class ControllerActionInvokerTest
...
         {
             // Arrange
             var filter1 = new Mock<IExceptionFilter>(MockBehavior.Strict);
 
+            var filter2 = new Mock<IExceptionFilter>(MockBehavior.Strict);
+            filter2
+                .Setup(f => f.OnException(It.IsAny<ExceptionContext>()))
+                .Callback<ExceptionContext>(context =>
+                {
+                    filter2.ToString();
```
Comment (rynowak) [inline test/Microsoft.AspNetCore.Mvc.Core.Test/Internal/ControllerActionInvokerTest.cs]: These calls to `ToString()` aren't needed. This is an **ancient** workaround for an issue in Moq. Just don't put them in new code

**Comment by rynowak** [comment]
:shipit: after removing those `ToString()`s

**Pull request merged**